### PR TITLE
[WIP] device dehydration

### PR DESCRIPTION
**Pull request by uhoreg**
implements https://github.com/matrix-org/matrix-doc/pull/2697

### CODE REVIEW

File: synapse/rest/client/v1/login.py [hunk 1]
```diff
@@ -401,6 +423,52 @@ async def _do_jwt_login(self, login_submission: JsonDict) -> Dict[str, str]:
         return result
 
 
+class RestoreDeviceServlet(RestServlet):
+    PATTERNS = client_patterns("/org.matrix.msc2697/restore_device")
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:427]: Not sure if this is supposed to show up only in `unstable` or also `r0` (the current code would do both).

File: synapse/rest/client/v1/login.py [hunk 2]
```diff
@@ -339,6 +340,27 @@ async def _complete_login(
...
+                )
+                result = {
+                    "user_id": user_id,
+                    "home_server": self.hs.hostname,
+                    "device_data": dehydrated_device,
+                    "device_id": device_id,
+                    "dehydration_token": token,
+                }
+
+                # FIXME: call callback?
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:360]: The [documentation about what this callback might be](https://github.com/matrix-org/synapse/blob/master/docs/password_auth_providers.md#optional-methods) is less than ideal. It comes from password auth modules and might be `None` or a `Callable` that is "called with the result from the `/login` call (including `access_token`, `device_id`, etc.)"

I do not know of any auth providers which use this frankly, but I think calling it like below is the reasonable thing to do. Is there a particular reason you think it should not be called?

Comment (clokep) [reply synapse/rest/client/v1/login.py:360]: We talked a bit more about this and the reason we can't just call it here is that we don't yet have the access token (the client hasn't acknowledged that they've been able to use the dehydrated device).

It seems like we would ideally want to call this at the end of the call to `/restore_device`, but we no longer have the callback at that point. Will need to think about what the best option is here.

**Comment by clokep** [comment]
I think we've decided the implementation in #8380 is a bit less invasive. Shall we close this?

**Comment by uhoreg** [comment]
replaced by 8380

**Pull request closed without merging**